src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs

```diff
@@ -0,0 +1,241 @@
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.AspNetCore.Http.Features;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic.Internal
+{
+    internal class MsQuicStream : TransportConnection
+    {
+        private Task _processingTask;
+        private QuicStream _stream;
+        private MsQuicConnection _connection;
+        private readonly CancellationTokenSource _streamClosedTokenSource = new CancellationTokenSource();
+
+        public MsQuicStream(MsQuicConnection connection, QUIC_STREAM_OPEN_FLAG flags)
+        {
+            var inputOptions = new PipeOptions(MemoryPool, PipeScheduler.ThreadPool, PipeScheduler.Inline, 1024 * 1024, 1024 * 1024 / 2, useSynchronizationContext: false);
+            var outputOptions = new PipeOptions(MemoryPool, PipeScheduler.Inline, PipeScheduler.ThreadPool, 1024 * 64, 1024 * 64 / 2, useSynchronizationContext: false);
+
+            var pair = DuplexPipe.CreateConnectionPair(inputOptions, outputOptions);
+            if (flags.HasFlag(QUIC_STREAM_OPEN_FLAG.UNIDIRECTIONAL))
+            {
+                Features.Set<IUnidirectionalStreamFeature>(new UnidirectionalStreamFeature());
+            }
+
+            // Add a fake TLS connection feature becuase quic is always secure
+            var feature = new FakeTlsConnectionFeature();
+            Features.Set<ITlsConnectionFeature>(feature);
+
+            // Set the transport and connection id
+            Transport = pair.Transport;
+            Application = pair.Application;
+            _connection = connection;
+            ConnectionClosed = _streamClosedTokenSource.Token;
+        }
+
+        public MsQuicStream(QuicStream stream, bool _)
+        {
+            var inputOptions = new PipeOptions(MemoryPool, PipeScheduler.ThreadPool, PipeScheduler.Inline, 1024 * 1024, 1024 * 1024 / 2, useSynchronizationContext: false);
+            var outputOptions = new PipeOptions(MemoryPool, PipeScheduler.Inline, PipeScheduler.ThreadPool, 1024 * 64, 1024 * 64 / 2, useSynchronizationContext: false);
+
+            var pair = DuplexPipe.CreateConnectionPair(inputOptions, outputOptions);
+
+            // Set the transport and connection id
+            Transport = pair.Transport;
+            Application = pair.Application;
+            _stream = stream;
+        }
+
+        // TODO make start async now that we have an event.
+        public void Start(QuicStream stream)
+        {
+            _stream = stream;
+            stream.SetCallbackHandler(HandleStreamEvent);
+            _processingTask = ProcessSends();
+        }
+
+        private async Task ProcessSends()
+        {
+            var output = Output;
+            try
+            {
+                while (true)
+                {
+                    var result = await output.ReadAsync();
+
+                    if (result.IsCanceled)
+                    {
+                        break;
```

Comment:
Is this graceful of abort?

Reply:
I don't think it can be canceled here.

Reply:
Actually, if I implement abort, it would only for abort. I'll do that now.